crates/iroha_core/src/state/storage_transactions.rs

```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    /// Create block to aggregate updates
+    pub fn block(&self) -> TransactionsBlock {
+        self.block_impl(false)
+    }
+
+    /// Create block to aggregate updates and revert changes created in the latest block
+    pub fn block_and_revert(&self) -> TransactionsBlock {
+        self.block_impl(true)
+    }
+
+    fn block_impl(&self, revert: bool) -> TransactionsBlock {
+        let guard = self.write_lock.lock();
+        TransactionsBlock {
+            latest_block_ref: &self.latest_block,
+            blocks_ref: &self.blocks,
+            _guard: guard,
+            revert,
+            current_block: None,
+        }
+    }
+}
+
+/// Persistent view of storage at certain point in time
+pub trait TransactionsReadOnly {
+    /// Read entry from the storage
+    fn get<Q>(&self, key: &Q) -> Option<Value>
+    where
+        Key: Borrow<Q>,
+        Q: Hash + Eq + ?Sized;
+}
+
+/// Module for [`TransactionsView`] and it's related impls
+mod view {
+    use super::*;
+
+    /// Consistent view of the storage at the certain version
+    pub struct TransactionsView<'storage> {
+        pub(super) latest_block: Option<Arc<BlockInfo>>,
+        /// Some transactions may be added to the map after `Self` is created,
+        /// but for us exists only transactions with `height < latest_block.height`
+        pub(super) blocks: &'storage DashMap<Key, Value>,
+    }
+
+    impl TransactionsReadOnly for TransactionsView<'_> {
+        fn get<Q>(&self, key: &Q) -> Option<Value>
+        where
+            Key: Borrow<Q>,
+            Q: Hash + Eq + ?Sized,
+        {
+            let Some(block) = &self.latest_block else {
+                return None;
+            };
+            let block_height = block.height;
+            if block.transactions.contains(key) {
+                return Some(block_height);
+            }
+
+            self.blocks
+                .get(key)
+                .map(|h| *h)
+                .filter(|&h| h < block_height)
+        }
+    }
+}
+pub use view::TransactionsView;
+
+/// Module for [`TransactionsBlock`] and it's related impls
+mod block {
+    use std::ops::Deref;
+
+    use super::*;
+
+    /// Batched update to the storage that can be reverted later
+    pub struct TransactionsBlock<'storage> {
+        /// References to [`TransactionsStorage`] struct
+        pub(super) latest_block_ref: &'storage ArcSwapOption<BlockInfo>,
+        pub(super) blocks_ref: &'storage DashMap<Key, Value>,
+        pub(super) _guard: MutexGuard<'storage, RawMutex, ()>,
+
+        /// Own fields
+        pub(super) revert: bool,
+        pub(super) current_block: Option<Arc<BlockInfo>>,
+    }
+
+    impl TransactionsBlock<'_> {
+        /// Add transactions of the block
+        pub fn insert_block(&mut self, transactions: HashSet<Key>, height: Value) {
+            let block_info = BlockInfo {
+                transactions,
+                height,
+            };
+            assert!(
```

Comment:
```suggestion
            debug_assert!(
```

Though, it isn't that much of a hot part. What do you think?

Reply:
I think we can keep `assert` here since the check is fast (`is_none`) and as you said, this function is not hot (called once per block)